File: __contest-2024-6-8/3.js
// 3180. Maximum Total Reward Using Operations I

function maxTotalReward(rewardValues) {
  rewardValues.sort((a, b) => a - b);
  let maximum = rewardValues[rewardValues.length - 1];

  let dp = new Array(maximum * 2).fill(0);

  for (let rew of rewardValues) {
    for (let i = 0; i < rew; i++) {
      if (dp[i + rew] < dp[i] + rew) {
        dp[i + rew] = dp[i] + rew;
      }
    }
  }

  return Math.max(...dp);
}
console.log(maxTotalReward([3, 7, 2, 9, 4]));

File: __contest-2024-6-8/3.rb
def max_total_reward(reward_values)
  reward_values.sort!
  maximum = reward_values[-1]
  dp = Array.new(maximum * 2, 0)

  reward_values.each do |rew|
    (0...rew).each do |i|
      if dp[i + rew] < dp[i] + rew
        dp[i + rew] = dp[i] + rew
      end
    end
  end

  dp.max
end

p max_total_reward([3, 7, 2, 9, 4])

# For every reward value less than `rew` , update DP
